backend: add Azure Blob Storage (ABS) backup support

vdice opened the pull request:
This PR adds Azure Blob Storage (ABS) backup support.

Accomplishes part of https://github.com/coreos/etcd-operator/issues/1110

[skip ci]

hongchaodeng commented:
Thanks for the contribution.
Can you discuss about your design (api, spec change) first in https://github.com/coreos/etcd-operator/issues/1110 ?

vdice commented:
@hongchaodeng design document added in latest push; please see `doc/design/abs_backup.md`.  Thank you!

xiang90 commented on doc/design/abs_backup.md:
````diff
@@ -0,0 +1,68 @@
+
+When starting etcd operator, we need to provide flags in order to access the ABS storage account:
+```bash
+$ etcd-operator --backup-abs-secret ${secret_name} --backup-abs-container ${container_name} ...
````
if both s3 and abs options are given, users can choose from s3 or abs? they are not mutual-exclusive, right?

vdice replied:
Correct.

xiang90 commented on pkg/backup/abs.go:
```diff
@@ -0,0 +1,103 @@
+	"io/ioutil"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/coreos/etcd-operator/pkg/backup/abs"
```
separate internal and external imports

xiang90 commented on pkg/backup/abs.go:
```diff
@@ -0,0 +1,103 @@
+func (ab *absBackend) getBlobSize(key string) (int64, error) {
+	rc, err := ab.open(key)
+
+	b, err := ioutil.ReadAll(rc)
```
do you need to close rc on error?

vdice replied:
I've updated to close rc on error and otherwise return `ioutil.Close(rc)` as the error returned from the function -- how does this sound?

xiang90 commented on pkg/backup/abs/abs.go:
```diff
@@ -0,0 +1,161 @@
+	containerExists, err := containerRef.Exists()
+	if err != nil {
+		return nil, err
+	} else if !containerExists {
```
change this to `if !containerExists`

xiang90 replied:
else is not needed.

xiang90 commented on pkg/backup/abs/abs.go:
```diff
@@ -0,0 +1,161 @@
+	opts := &storage.DeleteBlobOptions{}
+	err := blob.Delete(opts)
+
+	return err
```
return blob.Delete(&storage.DeleteBlobOptions{})

xiang90 commented on pkg/backup/abs/abs.go:
```diff
@@ -0,0 +1,161 @@
+	blob := w.container.GetBlobReference(blobName)
+
+	opts := &storage.GetBlobOptions{}
+	resp, err := blob.Get(opts)
```
return blob.Get(&storage.GetBlobOptions{})

xiang90 commented on pkg/backup/abs_backend_test.go:
```diff
@@ -0,0 +1,211 @@
+		t.Fatal(err)
+	}
+	if err.Error() != "missing required environment variable of AZURE_STORAGE_ACCOUNT" {
+		t.Fatal(err)
```
t.Error("explain the problem here")

xiang90 commented:
@vdice Looks good in general. @hongchaodeng needs to approve this.

ultimateboy commented on pkg/backup/abs/abs.go:
```diff
@@ -0,0 +1,161 @@
+		return fmt.Errorf("create block blob from reader failed: %v", err)
+	}
+
+	return err
```
this should probably just be `return nil`

hongchaodeng commented:
Code freeze right now. K8s 1.7 is released and it breaks a lot of our tests. We need to fix failures and make tests stable before merging any feature.

vdice commented:
PR updated/rebased from master; also added changelog addition.  PTAL @hongchaodeng

hongchaodeng commented:
We won't merge any feature for a while.
First, we need to make CRD migration happen. TPR will be removed in k8s 1.8 . For survival, we need  to put all our efforts into migration and won't have much time on other stuff.
Second, for longer term, we will move to more extensible backup storage. We will eventually move all stuff except PV over there too due to security concerns and easier management, e.g. separate persistent disk to save the credentials and config.

vdice commented:
@hongchaodeng rebased off master and tested to success on a (minikube) k8s 1.7.0 cluster -- please let me know anything looks awry, when convenient.  Thank you!

xiang90 commented on example/secret-abs-credentials.yaml.template:
```diff
@@ -0,0 +1,8 @@
+type: Opaque
+stringData:
+  storage-account: <storage-account-name>  
+  storage-key: <storage-key>
```
new line?

xiang90 commented on example/example-etcd-cluster-with-abs-backup.yaml:
```diff
@@ -0,0 +1,19 @@
+  ## the section below can be added to restore from a pre-existing cluster backup
+  restore:
+    backupClusterName: "cluster-a"
+    storageType: "ABS"
```
new line

xiang90 commented:
@hongchaodeng 

I feel this is OK to merge. I am OK with adding Azure/AWS/GCP support in tree initially. Then we need to figure out a way to add out of tree support for backup storage.

I give this pr a quick look. It seems OK. defer to @hongchaodeng, @hasbro17 and @fanminshi for a detailed review.

hongchaodeng commented:
Yeah. I agree that we should merge this and move forward.
After we finished the CRD migration, our focus has moved a bit off etcd operator and doesn't have much time making the extensible design for a short term. Let's merge this for easier cloud support at this moment.

hongchaodeng commented on cmd/operator/main.go:
```diff
@@ -76,6 +79,9 @@ func init() {
 	flag.StringVar(&awsConfig, "backup-aws-config", "",
 		"The name of the kube configmap object that stores the AWS config file. The file name must be 'config'.")
 	flag.StringVar(&s3Bucket, "backup-s3-bucket", "", "The name of the AWS S3 bucket to store backups in.")
+	flag.StringVar(&absSecret, "backup-abs-secret", "",
```
We will deprecate the the operator wide flag for backup storage options.
Each cluster's backup storage option should sit in its own cluster spec.

hongchaodeng commented on example/example-etcd-cluster-with-abs-backup.yaml:
```diff
@@ -0,0 +1,19 @@
+apiVersion: "etcd.database.coreos.com/v1beta2"
```
The example in spec_examples.md is good enough.
To keep it consistent, no duplicate example here again.

hongchaodeng commented on doc/user/backup_config.md:
```diff
@@ -128,3 +129,57 @@ spec:
+
+See the [ABS backup with cluster specific configuration](spec_examples.md#abs-backup-and-cluster-specific-abs-configuration) spec to see what the cluster's `spec.backup` field should be configured as to set a cluster specific ABS backup configuration. The following additional fields need to be set under the cluster spec's `spec.backup.abs` field:
+- `absContainer`: The name of the ABS container to store backups in.
+- `absSecret`: The secret object name which should contain two key value pairs for the ABS account name and key.
```
Please mention the file layout of the secret:

```
storage-account:
storage-key:
```

hongchaodeng commented on example/secret-abs-credentials.yaml.template:
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
```
put this under example/azure/ ?

hongchaodeng commented on glide.yaml:
```diff
@@ -25,3 +25,5 @@ import:
 - package: github.com/jpillora/go-ogle-analytics
 - package: golang.org/x/net
 - package: golang.org/x/time
+- package: github.com/Azure/azure-sdk-for-go/storage
```
Put glide.yaml, glide.lock change in separate commit ?

hongchaodeng replied:
It is no different to vendor the entire repo or specific dir under the repo.
Just 

```
package: github.com/Azure/azure-sdk-for-go
version ...
```

BTW, it's `v10.3.0-beta` now :)

hongchaodeng commented:
@vdice 
I would like to move forward this PR. Here's some thoughts right now:
- First, I would like this PR to be separated into four smaller PRs to better review it
  1. Separate the design docs, user docs, spec changes into a PR and submit first
  2. Separate the backup sidecar (mainly pkg/backup/backup.go) related changes into a PR and submit. 
  3. Finish the operator code.
  4. Add testing. Backup sidecar should be testable since it is just an HTTP endpoint. See *s3_backend_test.go* . I will help setup **azurite** container running in jenkins.
- The operator flag is not recommended. Just add the options in cluster spec. That's the way to go. We have moved storage class into cluster spec too. All backup storage related flags will be deprecated.

vdice commented:
@hongchaodeng Great news!  I'll get to work addressing code comments and breaking this up into separate PRs as requested.  Will have an update (and new PRs) by tomorrow.

vdice commented:
@hongchaodeng pushed an additional commit that hopefully addresses all feedback.  Wanted to make sure I'm on the right track prior to breaking these changes out.  

In particular, hoping for confirmation that I fully removed operator flag/backup customization in anticipation of the deprecation you've mentioned.  Thanks!

hongchaodeng commented on pkg/cluster/backupstorage/abs.go:
```diff
@@ -0,0 +1,90 @@
+		return err
+	}
+
+	os.Setenv(env.ABSStorageAccount, string(se.Data[spec.ABSStorageAccount]))
```
Don't set env. This will affect other clusters.

hongchaodeng commented on doc/user/backup_config.md:
```diff
@@ -4,6 +4,7 @@ In etcd operator, we provide the following options to save cluster backups to:
 - Persistent Volume (PV) on GCE or AWS
 - Persistent Volume (PV) with custom StorageClasses
 - S3 bucket on AWS
+- ABS container on Azure
```
Azure Blob Storage (ABS) container

vdice commented:
(work split up into PRs linked above -- follow those for progress)
